home-application/src/main/java/com/softserveinc/ita/homeproject/application/service/QueryApiService.java

```diff
@@ -0,0 +1,128 @@
+import com.softserveinc.ita.homeproject.homeservice.dto.BaseDto;
+import com.softserveinc.ita.homeproject.homeservice.exception.BadRequestHomeException;
+import com.softserveinc.ita.homeproject.homeservice.service.QueryableService;
+import org.springframework.data.domain.Page;
+import org.springframework.data.jpa.domain.Specification;
+
+/**
+ * QueryApiService - service that provides Spring Data Page by request query
+ *
+ * @author Oleksii Zinkevych
+ * @param <T> - entity type
+ * @param <D> - DTO type
+ * @see javax.ws.rs.core.UriInfo
+ */
+public interface QueryApiService<T extends BaseEntity, D extends BaseDto> {
+    Map<String, String> EXCLUDED_PARAMETERS = Arrays.stream(DefaultQueryParams.values())
+            .collect(Collectors.toMap(DefaultQueryParams::getParameter, DefaultQueryParams::getValue));
+
+    /**
+     *
+     * @param uriInfo - object that implements UriInfo interface
+     *                  that provides access to application and request URI information
+     * @return Map of query and path parameters required by EntitySpecificationService
+     * to build Filter specification
+     */
+    static MultivaluedMap<String, String> getFilterMap(UriInfo uriInfo) {
+        MultivaluedMap<String, String> queryParams = uriInfo.getQueryParameters();
+        queryParams.forEach(QueryApiService::validateQueryParamValue);
+        return Stream.of(uriInfo.getPathParameters(), queryParams)
+                .flatMap(map -> map.entrySet().stream())
+                .filter(not(entry -> EXCLUDED_PARAMETERS.containsKey(entry.getKey())))
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        Map.Entry::getValue,
+                        (v1, v2) -> {
+                            throw new BadRequestHomeException("Request contains same path and query params");
+                        },
+                        MultivaluedHashMap::new
+                ));
+    }
+
+    private static void validateQueryParamValue(String queryParamName, List<String> queryParamValue) {
+        if (queryParamValue.size() > 1) {
+            throw new BadRequestHomeException("Query param '" + queryParamName + "' has more than one value");
+        }
+    }
+
+    /**
+     *
+     * @param uriInfo - object that implements UriInfo interface
+     *                  that provides access to application and request URI information
+     * @return Spring Data Specification of entity
+     */
+    Specification<T> getSpecification(UriInfo uriInfo);
+
+    /**
+     *
+     * @param param   - String name of query parameter
+     * @param uriInfo - object that implements UriInfo interface
+     *                  that provides access to application and request URI information
+     * @return Optional of String value of specified query parameter
+     */
+    static Optional<String> getParameterValue(String param, UriInfo uriInfo) {
+        if (uriInfo.getQueryParameters().get(param) != null) {
```

Comment:
It could be applicable here but in general, try to keep your method to have a single entry and a single exit meaning only one return statement.

Together with Optional in your case could be achieved as the following:
```
return Optional.ofNullable(uriInfo.getQueryParameters().get(param))
    .map(<map to Iterable>)
    .map(<get next as first>);
```